src/ipc/ipc.c

```diff
@@ -310,12 +312,22 @@ int ipc_buffer_free(struct ipc *ipc, uint32_t buffer_id)
 		/* check comp state if sink and source are valid */
 		if (ibd->cb->sink == icd->cd &&
 		    ibd->cb->sink->state != COMP_STATE_READY)
-			return -EINVAL;
+			sink_state_invalid = true;
 		if (ibd->cb->source == icd->cd &&
 		    ibd->cb->source->state != COMP_STATE_READY)
-			return -EINVAL;
+			source_state_invalid = true;
 	}
 
+	/*
+	 * A buffer should only be prevented from being freed when both the
```

Comment:
This is difficult to follow, would it be easier to state when the buffer should be freed ?

Reply:
@lgirdwood sure, I will flip the logic. We're stil running some tests and once we fix the issues, I will update this PR.